File: server/models/diet_plan.py
import uuid
from datetime import datetime, timezone
import enum

from sqlalchemy import String, DateTime, ForeignKey, Text, Enum as SAEnum, Float
from sqlalchemy.orm import Mapped, mapped_column, relationship

from server.database.base import Base


def _uuid() -> str:
    return str(uuid.uuid4())


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class ActivityLevel(str, enum.Enum):
    LOW = "low"
    MODERATE = "moderate"
    HIGH = "high"


class BodyCondition(str, enum.Enum):
    UNDERWEIGHT = "underweight"
    IDEAL = "ideal"
    OVERWEIGHT = "overweight"
    UNKNOWN = "unknown"


class DietPlan(Base):
    __tablename__ = "diet_plans"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=_uuid)
    pet_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("pets.id", ondelete="CASCADE"), nullable=False, index=True
    )
    owner_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    activity_level: Mapped[ActivityLevel] = mapped_column(
        SAEnum(ActivityLevel, name="activity_level"), default=ActivityLevel.MODERATE, nullable=False
    )
    body_condition: Mapped[BodyCondition] = mapped_column(
        SAEnum(BodyCondition, name="body_condition"), default=BodyCondition.UNKNOWN, nullable=False
    )
    current_weight_kg: Mapped[float | None] = mapped_column(Float, nullable=True)
    goal_weight_kg: Mapped[float | None] = mapped_column(Float, nullable=True)
    allergies: Mapped[str | None] = mapped_column(String(500), nullable=True)
    health_conditions: Mapped[str | None] = mapped_column(String(500), nullable=True)
    additional_notes: Mapped[str | None] = mapped_column(String(1000), nullable=True)

    ai_recommendation: Mapped[str] = mapped_column(Text, nullable=False)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=_utcnow, nullable=False
    )

    pet = relationship("Pet", backref="diet_plans")

File: client/types/dietPlan.ts
export type ActivityLevel = "low" | "moderate" | "high";
export type BodyCondition = "underweight" | "ideal" | "overweight" | "unknown";

export interface DietPlan {
  id: string;
  petId: string;
  ownerId: string;
  activityLevel: ActivityLevel;
  bodyCondition: BodyCondition;
  currentWeightKg: number | null;
  goalWeightKg: number | null;
  allergies: string | null;
  healthConditions: string | null;
  additionalNotes: string | null;
  aiRecommendation: string;
  createdAt: string;
}

export interface DietPlanRequestPayload {
  petId: string;
  activityLevel: ActivityLevel;
  bodyCondition: BodyCondition;
  currentWeightKg?: number;
  goalWeightKg?: number;
  allergies?: string;
  healthConditions?: string;
  additionalNotes?: string;
}

interface RawDietPlan {
  id: string;
  pet_id: string;
  owner_id: string;
  activity_level: ActivityLevel;
  body_condition: BodyCondition;
  current_weight_kg: number | null;
  goal_weight_kg: number | null;
  allergies: string | null;
  health_conditions: string | null;
  additional_notes: string | null;
  ai_recommendation: string;
  created_at: string;
}

export function mapDietPlan(raw: RawDietPlan): DietPlan {
  return {
    id: raw.id,
    petId: raw.pet_id,
    ownerId: raw.owner_id,
    activityLevel: raw.activity_level,
    bodyCondition: raw.body_condition,
    currentWeightKg: raw.current_weight_kg,
    goalWeightKg: raw.goal_weight_kg,
    allergies: raw.allergies,
    healthConditions: raw.health_conditions,
    additionalNotes: raw.additional_notes,
    aiRecommendation: raw.ai_recommendation,
    createdAt: raw.created_at,
  };
}

export function toDietPlanPayload(payload: DietPlanRequestPayload) {
  return {
    pet_id: payload.petId,
    activity_level: payload.activityLevel,
    body_condition: payload.bodyCondition,
    current_weight_kg: payload.currentWeightKg ?? null,
    goal_weight_kg: payload.goalWeightKg ?? null,
    allergies: payload.allergies || null,
    health_conditions: payload.healthConditions || null,
    additional_notes: payload.additionalNotes || null,
  };
}
